nixos/nextcloud: add backup & restore scripts

Triton171 opened the pull request:
## Description of changes

Added backup & restore scripts to the `services.nextcloud` module. Since correctly backing up the entire state of an instance can be complicated and depends on the configuration of the instance (e.g. database config, data directory), I think it makes sense to include this in the module.

The scripts are mainly based on the [Nextcloud documentation](https://docs.nextcloud.com/server/latest/admin_manual/maintenance/backup.html). In order to make it easy to integrate it into different backup tools, the backup script doesn't do any compression, deduplication or versioning: Most people will probably want to use something like Restic or BorgBackup which does this anyways and otherwise one can simply call `tar` after `nextcloud-backup`.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/a-high-level-nixos-module-for-nextcloud/36975/4

RaitoBezarius reviewed (changes requested):
My 2 cents would be to create a `nextcloud-backup.nix` and extend the NextCloud module from there so that the maintenance of the backupping system is not pushed to the existing NextCloud maintainers.

For the testing, I think there should be something that checks for the availability of the backup option.

I need to think a bit more about this design to see how we can make this optimal for nixpkgs inclusion.

bjornfor commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -93,6 +93,176 @@ let
+    then
+      {
+        # The sqlite database is automatically backed up since it's contained in the datadir
+        backupFileName = "data/nextcloud.db";
```
`nextcloud` above is actually `${config.services.nextcloud.config.dbname}`.

bjornfor replied:
I know about that option because I came to NixOS nextcloud module from an old owncloud install -- which required me to set `dbname = "owncloud"`.

Triton171 replied:
Thanks, I fixed it in the latest version.

bjornfor reviewed (commented):
`/var/lib/redis-nextcloud` might also need backup, independently of cfg.config.dbtype. Nextcloud pulls in redis with e.g. `services.nextcloud.configureRedis = true`.

Triton171 commented:
> My 2 cents would be to create a nextcloud-backup.nix and extend the NextCloud module from there so that the maintenance of the backupping system is not pushed to the existing NextCloud maintainers.

That seems like a good idea, especially considering that the nextcloud module is already quite complex. Should I also add myself as a maintainer for the `nextcloud-backup` module, so that I get pinged on issues? I noticed that the nextcloud module doesn't set the `meta.maintainers` attribute.

> For the testing, I think there should be something that checks for the availability of the backup option.

Do you mean checking that `services.nextcloud.backupScript` is set by default to the script? The integration tests already check that `nextcloud-backup` and `nextcloud-restore` are available in the system path.

> /var/lib/redis-nextcloud might also need backup, independently of cfg.config.dbtype. Nextcloud pulls in redis with e.g. services.nextcloud.configureRedis = true.

Since redis is just used for caching, I don't think we need to back it up and the cache will simply be rebuilt after restoring the instance (the official docs also don't mention this).

bjornfor commented:
> > /var/lib/redis-nextcloud might also need backup, independently of cfg.config.dbtype. Nextcloud pulls in redis with e.g. services.nextcloud.configureRedis = true.
> 
> Since redis is just used for caching, I don't think we need to back it up and the cache will simply be rebuilt after restoring the instance (the official docs also don't mention this).

Maybe you're right. I was reading between the lines here (https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/files_locking_transactional.html), quoting:

> Memcached, the popular distributed memory caching system, is not suitable for the new file locking because it is not designed to store locks, and data can disappear from the cache at any time. Redis is a key-value store, and it guarantees that cached objects are available for as long as they are needed.

jorsn commented:
> > > /var/lib/redis-nextcloud might also need backup, independently of cfg.config.dbtype. Nextcloud pulls in redis with e.g. services.nextcloud.configureRedis = true.
> > 
> > 
> > Since redis is just used for caching, I don't think we need to back it up and the cache will simply be rebuilt after restoring the instance (the official docs also don't mention this).
> 
> Maybe you're right. I was reading between the lines here (https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/files_locking_transactional.html), quoting:
> 
> > Memcached, the popular distributed memory caching system, is not suitable for the new file locking because it is not designed to store locks, and data can disappear from the cache at any time. Redis is a key-value store, and it guarantees that cached objects are available for as long as they are needed.

While backing up/restoring, Nextcloud is in maintenance mode anyway, so there shouldn't be any locks in the redis cache. The remaining cache can be rebuilt.

Andreas02-dev commented:
What is needed to complete this work, aside from the merge conflicts?
I'd love to see this added to the Nextcloud module.

MartiniMoe commented:
Hi,

I'm currently trying to implement my Nextcloud backups with a similar script, but somewhat different, because I directly want to backup the data directory with borg backup.

The problem I'm facing is, that my nextcloud cannot be set to maintenance mode, even when I run the borg backup as the nextcloud user:

```
Jun 08 16:25:11 myhost borgbackup-job-nextcloud_data-start[112532]: Cannot write into "config" directory!
Jun 08 16:25:11 myhost borgbackup-job-nextcloud_data-start[112532]: This can usually be fixed by giving the web server write access to the config directory.
Jun 08 16:25:11 myhost borgbackup-job-nextcloud_data-start[112532]: But, if you prefer to keep config.php file read only, set the option "config_is_read_only" to true in it.
Jun 08 16:25:11 myhost borgbackup-job-nextcloud_data-start[112532]: See https://docs.nextcloud.com/server/29/go.php?to=admin-config
```

Has anyone of you faced this issue? Or do you have any ideas what else might be needed so successfully set maintenance mode from my script? Thanks :)

Triton171 commented:
Sorry for taking so long to respond, I was a bit busy. I rebased onto the latest `master`, which mainly required fixing the tests (the only thing I changed in the actual scripts is adding the `--if-exists` flag to `pg_restore`). The PR should be done now, so the only thing missing is a maintainer willing to review it. I'm new to contributing to NixOS so please let me know if there's anything else I should do to hopefully get this merged.

Oudwins commented:
Any progress on this? Seems like it has been quite a few months still no one had the time to review it? :(

Ma27 commented:
I wasn't even aware of this PR. Hopefully, I'll be able to take a look at it soonish, but this week I seem to be drowning in other stuff, so there are a bunch of other things awaiting review.

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+      fi
+    done
+  '';
+  backupScript = pkgs.writeScriptBin "nextcloud-backup" ''
+    #! ${pkgs.runtimeShell}
```
```suggestion
  backupScript = pkgs.writeShellScriptBin "nextcloud-backup" ''
```

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+    fi
+    # See https://docs.nextcloud.com/server/latest/admin_manual/maintenance/backup.html
+    echo "Enabling maintenance mode..."
+    ${cfg.occ}/bin/nextcloud-occ maintenance:mode --on
```
```suggestion
    ${lib.getExe cfg.occ} maintenance:mode --on
```

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+    ${checkBackup}
+    echo "Nextcloud backup completed."
+  '';
+  restoreScript = pkgs.writeScriptBin "nextcloud-restore" ''
+    #! ${pkgs.runtimeShell}
```
```suggestion
  restoreScript = pkgs.writeShellScriptBin "nextcloud-restore" ''
```

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+    NEXTCLOUD_MAJOR_VERSION=$(grep "^Nextcloud Version:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Nextcloud Version: ([0-9]+)\..*/\1/")
+    NEXTCLOUD_DB_TYPE=$(grep "^Database Type:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Database Type: //")
+    BACKUP_FORMAT_VERSION=$(grep "^Backup Format Version:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Backup Format Version: //")
+    # A few sanity checks to make sure that it's possible to restore the backup.
+    if [ ! "$BACKUP_FORMAT_VERSION" = "1" ] ;then
+      # In case there are backwards-incompatible changes to this script in the future
+      echo "Unknown backup format version $BACKUP_FORMAT_VERSION, aborting."
+      exit 1
+    fi
```
I think this might be a bit over engineered for this backup script

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -202,6 +202,7 @@ let
 in {
 
   imports = [
+    ./nextcloud-backup.nix
```
This should be placed in modules list

Ma27 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+      exit 1
+    fi
+    # Read backup metadata
+    NEXTCLOUD_MAJOR_VERSION=$(grep "^Nextcloud Version:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Nextcloud Version: ([0-9]+)\..*/\1/")
```
Wouldn't it make sense to use some kind of structured data here? Grepping for texts seems a little brittle IMHO.

Ma27 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+  '';
+in
+{
+  options.services.nextcloud = {
```
I must say, I'm not comfortable with having a non-official backup solution in the `services.nextcloud` namespace.

Would it be OK for everyone to move this into `services.nextcloudBackup` or something like this?
I think `mysql`/`postgresql` are good precedents for that.

Ma27 commented on nixos/tests/nextcloud/with-declarative-redis-and-secrets.nix:
```diff
@@ -122,5 +122,19 @@ in {
     nextcloud.fail('test "[]" = "$(redis-cli --json KEYS "*")"')
+
+    # Create a backup, nuke the instance, restore the backup & check that everything works
+    nextcloud.systemctl("start --wait nextcloud-update-db.service")
```
Hmm I don't know why this isn't merged with the other tests :thinking: 

Gonna check later, but nothing we should do in this PR.

Ma27 reviewed (commented):
I guess I'll be upfront and say that I'd like to have this separated (see comment below) and I don't intend to help maintaining this.

However, I do see a value in this and I think it's a good addition. Thank you for your work!

I went through the entire change and everything looks reasonable so far.
Just a few minor notes from my side.

Triton171 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+            export PGPASSWORD="$(cat ${cfg.config.dbpassFile})"
+          ''
+          else "";
+        commonFlags = "-h ${cfg.config.dbhost} -U ${cfg.config.dbuser} --no-password";
```
The `postgresqlBackup` module doesn't allow for multiple different backups. So if someone uses it to back up some databases (other than the nextcloud db), their backup would automatically include the nextcloud database as well since we'd have to include it in `services.postgresqlBackup.databases`. Also, backing up nextcloud would force a backup of all the other databases. I don't really see a way to avoid this conflict with the current `postgresqlBackup` module.

Triton171 commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,199 @@
+    else if cfg.config.dbtype == "mysql"
+    then
+      let
+        backupFileName = "database-mysql.bak";
+        mysqldump = "${config.services.mysql.package}/bin/mysqldump";
+        mysql = "${config.services.mysql.package}/bin/mysql";
+        # MySQL doesn't accept the socket path as a hostname
+        hostFlag = if cfg.database.createLocally then "-h localhost" else "-h ${cfg.config.dbhost}";
+        passwordFlag =
+          if cfg.config.dbpassFile != null
+          then ''
+            -p "$(cat ${cfg.config.dbpassFile})"
+          ''
+          else "";
+        commonFlags = hostFlag + " -u ${cfg.config.dbuser} " + passwordFlag;
```
This has the same issue as the `postgresqlBackup` module.

Triton171 committed:
nixos/nextcloudBackup: init module

Triton171 commented:
Thanks a lot for the reviews, i finally found some time to work on them. The latest commit should implement all the suggestions except for the ones regarding `services.postgresqlBackup` / `services.mysqlBackup` (see comment above). If i missed anything, let me know.

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,196 @@
+  dbBackupRestore =
+    if cfg.config.dbtype == "sqlite" then
+      {
+        # The sqlite database is automatically backed up since it's contained in the datadir
```
I think even though the database is contained int the datadir you should still export it as suggested in https://docs.nextcloud.com/server/latest/admin_manual/maintenance/backup.html. I'm not sure maintenance mode guarantees the database will not be touched. Additionally sqlites usage of a WAL might make the backup inconsistent.

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,196 @@
+        pg_dump = "${config.services.postgresql.package}/bin/pg_dump";
+        pg_restore = "${config.services.postgresql.package}/bin/pg_restore";
+        setPassword =
+          if cfg.config.dbpassFile != null then
```
You can simplify this:
```suggestion
          lib.optionalString (cfg.config.dbpassFile != null) "export PGPASSWORD=..."
```

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,196 @@
+      default = restoreScript;
+      defaultText = lib.literalMD "generated script";
+      internal = true;
+      description = lib.mdDoc ''
```
`lib.mdDoc` is deprecated as descriptions are markdown by default now.

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,196 @@
+      '';
+    };
+  };
+  config = lib.mkIf cfg.enable {
```
I think this should be a separate enable options because not everyone might need these scripts.

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,196 @@
+  };
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = [
+      backupScript
```
This doesn't use the options defined above so if I were to change them it wouldn't reflect here. Is this intentional, the why do the options even exist?

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,196 @@
+      type = lib.types.package;
+      default = backupScript;
+      defaultText = lib.literalMD "generated script";
+      internal = true;
```
Why are these internal? As far as I can see their not used anywhere so if they're supposed to be internal and not used by anyone why do they exist? On the contrary I think we should expose them if people want to write backup services, they should be able to access the scripts.

Triton171 replied:
I set them to `internal` because their not really meant to be set by the user. These options only exist to expose the scripts to the NixOS config (e.g. to run them in  `services.restic.backups.<name>.backupPrepareCommand`). I removed the `internal` flag and instead explained this in the description.

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup.nix:
```diff
@@ -0,0 +1,196 @@
+  backupScript = pkgs.writeShellScriptBin "nextcloud-backup" ''
+    set -Eeuo pipefail
+    if [ "$#" != "1" ] || [ "$1" = "-h" ] || [ "$1" = "--help" ] ;then
+      echo "Usage: nextcloud-backup TARGET_DIR"
```
It feels like this script is somewhere between a service and a package. It exists as a service, but then some options, like the `TARGET_DIR` aren't settable using nixos but need to be given on the command line. I think you should embrace the service part more, make all inputs configurable as nixos options, and maybe allow overrides on the command line. This would make it easier to use the script inside other modules, and might allow us for example to add our own timed service for backups at a later point.

PatrickDaG reviewed (changes requested):
Lastly the script has gotten a bit convoluted and hard to read. It might be good idea to move the shell part into an actual `backup.sh` to make it easier to see whats going on, similar to how [vaultwarden](https://github.com/NixOS/nixpkgs/blob/39495b6479aa8f8c132281090899ed0ea09d01bf/nixos/modules/services/security/vaultwarden/default.nix) does it.
Sorry for the large mount of nitpicking. All in all it looks great.

Triton171 commented:
No worries about nitpicking, I'm happy for all the constructive feedback (it just sometimes takes me a few weeks to find time to work on it :smile:). The latest version should address the remaining suggestions.

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup/backup.sh:
```diff
@@ -0,0 +1,72 @@
+# Nextcloud backup script, based on https://docs.nextcloud.com/server/latest/admin_manual/maintenance/backup.html.
+set -Eeuo pipefail
```
`writeShellApplication` sets `-e` `-u` and `pipefail`.
You can set `-E` by adding `errtrace` to the `bashOptions` for `writeShellApplication`

PatrickDaG commented on nixos/modules/services/web-apps/nextcloud-backup/default.nix:
```diff
@@ -0,0 +1,113 @@
+    name = "nextcloud-restore";
+    runtimeInputs = runtimeInputs;
+    text = ''
+      ${setEnvVariables}
```
You can set the env using the `runtimeEnv` argument and then set the text to the `.sh` file directly.

PatrickDaG reviewed (approved):
Just suggestions left, not really blockers. Hopefully gets merged soon.

Triton171 commented:
Thanks, that makes the code a lot nicer (and also now the shell-check tests of `writeShellApplication` are actually checking the scripts :D).

jockee commented:
Is this going to be merged?

Triton171 committed:
maintainers: add triton

Triton171 commented:
I rebased on top of master and made a few changes:
- The backup script is no longer exposed directly, but instead as a systemd service. This was necessary because `nextcloud-occ` has been changed to require either root permissions or a specific systemd service configuration. The restore script hasn't been changed (in my opinion, it's fine for this to require root permissions).
- Added some documentation for the backup module.
- Added myself as a maintainer for the backup module.

Otherwise, the code is essentially unchanged (except for some formatting changes). I'd appreciate any maintainers taking some time to review the latest version and hopefully get it merged. If there is any reason why you'd rather not merge (in its current state), please let me know!
